### PR TITLE
Add HashDirectiveInfo module to FCS.

### PR DESCRIPTION
**Pull request by smoothdeveloper**
Add HashDirectiveInfo module to FCS, this is used to implement "go to definition" over #load directives in script.

This is the editor agnostic code I originally made in VFPT, I'm not sure if I'll have time to integrate it in VS, but it will be useful for other editors.

Comes with a simple unit test.

### CODE REVIEW

File: src/fsharp/service/HashDirectiveInfo.fsi [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.FSharp.Compiler.SourceCodeServices
+
+#if COMPILER_PUBLIC_API
```
Comment (cartermp) [inline src/fsharp/service/HashDirectiveInfo.fsi]: Let's not worry about this given #4120 was merged

**Comment by cartermp** [comment]
I the intention here that I can have the following script:

```fsharp
#load "MyMath.fs"

open MyMath

square 12
```

And f12 into the `"MyMath.fs"` part of the script to take me to that file, once VS integration (and other integration) is added?

**Comment by smoothdeveloper** [comment]
@cartermp yes, the current code has two functions:

* list all #I and #load directives, with logic to resolve the absolute path
* check if such directives are at a specific position

I'm looking at the VS integration, and might have that in a separate PR eventually, but the logic could be useful already if say Ionide would like to implement the navigation feature.

**Comment by cartermp** [comment]
How do you feel about this versus the existing [FileSystemCompletion](https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs) logic that @vasily-kirichenko ported over from Roslyn?

I'm all for exposing this functionality in FCS so other editors can benefit, but we'll want to make sure we don't have two implementations here.

**Comment by smoothdeveloper** [comment]
I haven't looked into the implementation much but it is not based on the AST but rather it uses the roslyn toolset.

I think it makes sense to have the HashDirectiveModule (or equivalent) as part of FCS.

We can look if there is use to change  FileSystemCompletion to use HashDirectiveModule but I think the implementation choice was made this way to keep it close to C# version as well as performance reasons.

**Comment by cartermp** [comment]
Right, the actual searching logic is done in Roslyn, but there are still some differences here (e.g., handling network paths). Using the AST vs. a regex (what we have currently) to snag directives seems like it should really just be a matter of efficiency. Whichever is faster should be what's taken and placed in FCS.

In terms of the actual searching logic, how does it stack up when compared with [what Roslyn does](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Completion/FileSystemCompletionHelper.cs,118)?

**Comment by smoothdeveloper** [comment]
@cartermp / @KevinRansom I think this is ready, somehow forgot to ping about the requested change to remove `COMPILER_PUBLIC_API` which is now done.

Is there anyone which can point me how to open arbitrary file in Visual Studio editor, I'll want to integrate it in https://github.com/Microsoft/visualfsharp/blob/5a12cde14c18646380a05371297e38e74e96bc18/vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs but I couldn't figure out appropriate way to open editor given arbitrary file path from there last time I had a look.

**Comment by majocha** [comment]
@smoothdeveloper [VsShellUtilities.OpenDocument ](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.vsshellutilities.opendocument?view=visualstudiosdk-2017#Microsoft_VisualStudio_Shell_VsShellUtilities_OpenDocument_System_IServiceProvider_System_String_) maybe this will do?

**Comment by forki** [comment]
ping

**Comment by smoothdeveloper** [comment]
@forki I had issues building the repository: https://github.com/Microsoft/visualfsharp/pull/4317#issuecomment-377328919 I'll give it another try (thanks @majocha going to try this soon).

The FCS part is ready for merge though, I'll do the VS tooling part in a separate PR.

**Comment by KevinRansom** [comment]
@cartermp, 

can you work with @smoothdeveloper to see if you still need changes, thanks.

Kevin

**Comment by smoothdeveloper** [comment]
@KevinRansom thanks for the nudge, there are two points I wanted to make:

* other editors (ionide & rider AFAIR) seem to have implemented their own "go to definition" on `#load` directives, without use of this code, VS2017 is the only one not having it
* I've recently learned about `#load` for multiples files in a single directive recently and this code (which comes from VFPT) doesn't work properly with it, I think this is minor issue for now

I'm currently not able to dig back into the VS integration bits I experimented, but it would be great if someone at MS could give a shot at wiring the logic in this PR to the goto definition service using Roslyn APIs.

I currently can't commit the time fiddling with it but I'm sure that for someone who is sufficently exposed to APIs, it won't be too much trouble, the code here handles you an absolute path to open an editor for and the goto definition service isn't too big.

**Comment by KevinRansom** [comment]
@smoothdeveloper thanks for the heads up, and for the PR.   I will chat with a few folks to see how to take this forward.

Kevin

File: fcs/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj [hunk 2]
```diff
@@ -59,6 +59,17 @@
     <Compile Include="$(FSharpSourcesRoot)\..\tests\service\StructureTests.fs">
       <Link>StructureTests.fs</Link>
     </Compile>
+    <Compile Include="$(FSharpSourcesRoot)\..\tests\service\HashDirectiveInfoTests.fs">
+      <Link>HashDirectiveInfoTests.fs</Link>
+    </Compile>
+    <None Include="$(FSharpSourcesRoot)\..\packages\FSharp.Compiler.Tools.4.1.27\tools\FSharp.Core.optdata">
```
Comment (dsyme) [inline fcs/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj]: These should not be needed

File: src/fsharp/service/HashDirectiveInfo.fs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        // the Load items are resolved using fallback resolution relying on previously parsed #I directives
+        // (this behaviour is undocumented in F# but it seems to be how it works).
+
+        // list of #I directives so far (populated while encountering those in order)
+        // TODO: replace by List.fold if possible
+        let includesSoFar = new System.Collections.Generic.List<_>()
+        let pushInclude = includesSoFar.Add
+
+        // those might need to be abstracted away from real filesystem operations
+        let fileExists = File.Exists
```
Comment (dsyme) [inline src/fsharp/service/HashDirectiveInfo.fs]: Please remove all these abbreviations

Comment (dsyme) [reply src/fsharp/service/HashDirectiveInfo.fs]: You should also use the FileSystem abstraction for all of these

File: src/fsharp/service/HashDirectiveInfo.fsi [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    open Microsoft.FSharp.Compiler.Ast
+
+    /// IncludeDirective (#I) contains the pointed directory
+    type IncludeDirective =
+        | ResolvedDirectory of string
+    
+    /// a LoadDirective (#load) either points to an existing file, or an unresolveable location (along with all locations searched)
+    type LoadDirective =
+        | ExistingFile of string
+        | UnresolvableFile of string * previousIncludes : string array
```
Comment (dsyme) [inline src/fsharp/service/HashDirectiveInfo.fsi]: Please use `string[]` rather than `string array` in the compiler

**Comment by dsyme** [comment]
This looks like a partial reimplementation of aspects of the resolution logic done by the main F# compiler code.

The normal approach to this would be to _record_ the resolutions in from the type-checking/analysis phase and _report_ those resolutions here, rather than reimplementing the resolution logic.

In this case it's not a big problem.  But there's a bit of trend toward reimplementing core compiler logic under `src/fsharp/service` and of course in the long term that will be a maintenance problem.

**Comment by smoothdeveloper** [comment]
I'm resuming this in a new branch (https://github.com/smoothdeveloper/visualfsharp/tree/hashdirective-redux) and will open a separate PR.

Taking care of some reviewed aspects:

@dsyme

Ok I found the `FileSystem` shims and extending them, removing the custom path stuff; is it ok to use `System.IO.Path.Combine` in compiler or I should add `FileSystem.CombineShim` as well?

> This looks like a partial reimplementation of aspects of the resolution logic done by the main F# compiler code.
>
> The normal approach to this would be to record the resolutions in from the type-checking/analysis phase and report those resolutions here, rather than reimplementing the resolution logic.

Do you know the location of that resolution in the code? I'd be happy to take a further look at some point.

I'll add some more tests and figure out handling of multi file load directives, I'm not sure I can work yet on the VS bits in my environment.

**Comment by dsyme** [comment]
> Ok I found the FileSystem shims and extending them, removing the custom path stuff; is it ok to use System.IO.Path.Combine in compiler or I should add FileSystem.CombineShim as well?

It's ok to use `SYstem.IO.Path.Combine`

**Pull request closed without merging**